From e7958a259307bea8830bd274bc006313120eb315 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 19 Aug 2025 14:43:41 +0800
Subject: [PATCH 206/254] mtk: hostapd: make sure that
 hostapd_ml_get_assoc_sta() correctly returns

The function should reutrn the sta_info from it's setup link
(ie. sta->mld_assoc_sta == sta); otherwise, it should return NULL.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ieee802_11_shared.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
index 781374bc2..89d416df8 100644
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -1221,14 +1221,15 @@ struct sta_info * hostapd_ml_get_assoc_sta(struct hostapd_data *hapd,
 	*assoc_hapd = hapd;
 
 	/* The station is the one on which the association was performed */
-	if (sta->mld_assoc_link_id == hapd->mld_link_id)
+	if (sta->mld_assoc_link_id == hapd->mld_link_id &&
+	    sta == sta->mld_assoc_sta)
 		return sta;
 
 	other_hapd = hostapd_mld_get_link_bss(hapd, sta->mld_assoc_link_id);
 	if (!other_hapd) {
 		wpa_printf(MSG_DEBUG, "MLD: No link match for link_id=%u",
 			   sta->mld_assoc_link_id);
-		return sta;
+		return NULL;
 	}
 
 	/*
@@ -1237,14 +1238,14 @@ struct sta_info * hostapd_ml_get_assoc_sta(struct hostapd_data *hapd,
 	 */
 	for (tmp_sta = other_hapd->sta_list; tmp_sta; tmp_sta = tmp_sta->next) {
 		if (tmp_sta->mld_assoc_link_id == sta->mld_assoc_link_id &&
-		    tmp_sta->aid == sta->aid) {
+		    tmp_sta->aid == sta->aid && tmp_sta->mld_assoc_sta == tmp_sta) {
 			*assoc_hapd = other_hapd;
 			return tmp_sta;
 		}
 	}
 #endif /* CONFIG_IEEE80211BE */
 
-	return sta;
+	return NULL;
 }
 
 
-- 
2.45.2

